src/wallet/wallet.cpp

```diff
@@ -5015,6 +5035,42 @@ bool PaymentAddressBelongsToWallet::operator()(const libzcash::InvalidEncoding&
     return false;
 }
 
+boost::optional<libzcash::ViewingKey> GetViewingKeyForPaymentAddress::operator()(
+    const libzcash::SproutPaymentAddress &zaddr) const
+{
+    libzcash::SproutViewingKey vk;
+    if (!m_wallet->GetSproutViewingKey(zaddr, vk)) {
+        libzcash::SproutSpendingKey k;
+        if (!m_wallet->GetSproutSpendingKey(zaddr, k)) {
+            return boost::none;
+        }
+        vk = k.viewing_key();
+    }
+    return libzcash::ViewingKey(vk);
+}
+
+boost::optional<libzcash::ViewingKey> GetViewingKeyForPaymentAddress::operator()(
+    const libzcash::SaplingPaymentAddress &zaddr) const
+{
+    libzcash::SaplingIncomingViewingKey ivk;
+    libzcash::SaplingExtendedFullViewingKey extfvk;
+
+    if (m_wallet->GetSaplingIncomingViewingKey(zaddr, ivk) &&
+        m_wallet->GetSaplingFullViewingKey(ivk, extfvk))
+    {
+        return libzcash::ViewingKey(extfvk);
+    } else {
+        return boost::none;
+    }
+}
```

Comment:
The logic here is different from Sprout (besides having the intermediate ivk), in that if the wallet doesn't hold a Sprout viewing key but does hold a spending key, the viewing key will be derived from the spending key by `GetViewingKeyForPaymentAddress`. For Sapling I think we rely on `GetSaplingIncomingViewingKey` doing that derivation, am I correct?

Why the difference?

Reply:
Historical artifact. The underlying keystore stores a map from Sprout addresses to Sprout spending keys, and when we added viewing key support, we did so with a separate map from Sprout addresses to Sprout viewing keys instead of inserting it into the existing pathway. Sapling's more diverse key structure meant that we created a hierarchy internally, so there is already a pathway from Sapling addresses to spending keys that goes through viewing keys. We now just instantiate some of those from disk instead of purely from memory.